### PR TITLE
chore(BUX-457): Close response.Body & test with httpmock & fix major bug related with strategies

### PR DESCRIPTION
**Pull request by chris-4chain**
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/go-broadcast-client/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/go-broadcast-client/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [x] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

The initial task was to add `defer resp.Body.Close()` to `DoRequest` function
- To do that, I proposed a generic `RequestModel` function which should handle all http-request boilerplate along with proper response-body-close, error handling and parsing a response model.

After that, I wanted to test it properly, that's why I added `httpmock` library. 

Then, while adjusting tests I found a bug:

### The bug
It was standard [Golang trap](https://golang50shad.es/#anameclosure_for_it_varsaiterationvariablesandclosuresinforstatements) with Iteration Variables and Closures in "for" Statements.

The effect was that:
Suppose we have defined three miners [Gorilla, XYZ, Taal]. The buggy `oneByone` strategy tried to as three times the last element which is in this case Taal. 

After my changes it's fixed - Check tests which can prove that.

### CODE REVIEW

File: broadcast/errors.go [hunk 1]
```diff
@@ -21,9 +21,6 @@ var ErrClientUndefined = errors.New("client is undefined")
 // ErrAllBroadcastersFailed is returned when all configured broadcasters failed to query or broadcast the transaction.
 var ErrAllBroadcastersFailed = errors.New("all broadcasters failed")
 
-// ErrURLEmpty is returned when the API URL is empty.
-var ErrURLEmpty = errors.New("url is empty")
-
```
Comment (dorzepowski) [inline broadcast/errors.go:26]: Why did you remove it?

Comment (chris-4chain) [reply broadcast/errors.go:26]: I moved it to the http_client to release the client from `broadcast` package dependency.

File: httpclient/http_client.go [hunk 2]
```diff
@@ -95,18 +107,46 @@ func (hc *HTTPClient) DoRequest(ctx context.Context, pld HTTPRequest) (*http.Res
...
-	return nil, HttpClientError{resp}
-}
+	model, err = parser(resp)
+	if err != nil {
+		return
+	}
 
-func hasSuccessCode(resp *http.Response) bool {
-	return resp.StatusCode >= 200 && resp.StatusCode < 300
+	return
```
Comment (kuba-4chain) [inline httpclient/http_client.go:151]: It's a minor, but I'm not in favour of naked returns. 

I'm not gonna force it tho, just sayin. Unless someone agrees with me? 😄

File: httpclient/http_client.go [hunk 3]
```diff
@@ -95,18 +107,46 @@ func (hc *HTTPClient) DoRequest(ctx context.Context, pld HTTPRequest) (*http.Res
...
+type ParserFunction[T any] func(resp *http.Response) (T, error)
+type ErrorParserFunction func(statusCode int, body []byte) error
+
+func RequestModel[T any](
+	ctx context.Context,
+	req RequestFuncion,
+	pld HTTPRequest,
+	parser ParserFunction[T],
+	errorParser ErrorParserFunction,
+) (model T, err error) {
```
Comment (kuba-4chain) [inline httpclient/http_client.go:127]: When I saw this at first, I was sceptical. I think this is a little bit of an overkill to make such an abstraction, especially because the return type is steered by the parser function and is not known when looking at it at the first glance, which makes it really annoying.

But then I played with it a little bit and it turns out that IDE fills in the template type T with your parser return type and makes it much less annoying and I think it's really smart overall! If the team likes this approach, we could use it also in other libs where we have our own http client :)

File: httpclient/http_client.go (hunk 2 above)
Comment (chris-4chain) [reply httpclient/http_client.go:151]: Generally, I try to avoid naked returns. However, there is one exception when 'standard' returns can't be applied: it's the case (this case) when I want to return an error based on deferred recovery.

There might be another question: why did I decide to write the 'defer recover' part? 
There are two factors:

- This function makes an HTTP call to a remote server, and during that, unpredictable events might occur (such as timeouts or socket closures), so it's better to 'try-catch' potential errors.
- This function utilizes other functions (`parser` and `errorParser`) passed as arguments, so I can't assume that they will work without failures."

File: broadcast/internal/acceptance_tests/arc_fee_quote_test.go [hunk 4]
```diff
@@ -2,147 +2,103 @@ package acceptancetests
...
-	"github.com/rs/zerolog"
-	"io"
 	"net/http"
-	"strings"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
-	"github.com/stretchr/testify/mock"
+
+	"github.com/jarcoal/httpmock"
```
[the remark this answers is not in the document]
Comment (chris-4chain) [reply broadcast/internal/acceptance_tests/arc_fee_quote_test.go:10]: Because of I focused on `http_client` (RequestModel) function with `DoReqest` I tried to test it properly. 
I realized that, these functions/utilities are not properly tested in isolation. Furthermore, the other tests like `acceptance_tests` uses `testify/mock` which do not consider `DoRequest` part and the nature of actual http requests.

That's why I decided to introduce `httpmock` which I find very usefull. Thank's to this lib and changes I made, I found the quite serious bug I described in this PR. 

There is another issue when using `testify/mock` for http-request-related functions. 
It strongly depends on a developer who mocks the `DoRequest` functions whether a test passes or not. And it's more likely that, although tests passes, the actual request during runtime fails. I tried to avoid such scenarios by adding the `httpmock` lib.

File: httpclient/http_client.go (hunk 3 above)
Comment (chris-4chain) [reply httpclient/http_client.go:127]: I'm glad that you find this function useful. To be clear, I did not introduce this just because I like to write complicated generic constructs which contain all types of brackets. 

The main purpose of this task was to ensure that `resp.Body` entities are closed properly, so I had to find out how to do it once, and avoid the situation when the next developer "forgets" to close it. 
- First of all, I didn't want to modify public interfaces and the `DoRequest` function type. 
- Then, I realized that I couldn't close the body before a model is parsed. 
- Moreover, I found out that errors (especially `arcErrors`) are parsed based on the response body. 

Summing these factors up I figured, that I have to write a function which contains all the three parts `DoRequest`, `modelParsing` and `errorParsing` in one function and use a defer to properly close the response body at the end. 
That's the reason why I decided to do it this way :)

Comment (kuba-4chain) [reply httpclient/http_client.go:127]: Got it, that's another reason why we should the same in our other libs :)

**Review by kuba-4chain — APPROVED** [review]
